File: Editorial/102_Binary_Tree_Level_Order_Traversal/TypeScript/2_Iteration.ts
const levelOrder = function (root: TreeNode | null): number[][] {
    let levels: number[][] = [];
    if (!root) return levels;
    let queue: TreeNode[] = [root];
    let level: number = 0;
    while (queue.length) {
        // Start the current level
        levels.push([]);
        // Number of elements in the current level
        let level_length: number = queue.length;
        for (let i = 0; i < level_length; ++i) {
            let node: TreeNode | undefined = queue.shift();
            // Fulfill the current level
            if (node) levels[level].push(node.val);
            // Add child nodes of the current level
            // In the queue for the next level
            if (node?.left) queue.push(node.left);
            if (node?.right) queue.push(node.right);
        }
        // Go to next level
        level++;
    }
    return levels;
};

File: Editorial/0101-0200/0102_Binary_Tree_Level_Order_Traversal/CPP/2_Iteration.cpp
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> levels;
        if (root == NULL) return levels;
        deque<TreeNode*> queue;
        queue.push_back(root);
        int level = 0;
        while (!queue.empty()) {
            // start the current level
            levels.push_back({});
            // number of elements in the current level
            int level_length = queue.size();
            for (int i = 0; i < level_length; ++i) {
                TreeNode* node = queue.front();
                queue.pop_front();
                // fulfill the current level
                levels[level].push_back(node->val);
                // add child nodes of the current level
                // in the queue for the next level
                if (node->left != NULL) queue.push_back(node->left);
                if (node->right != NULL) queue.push_back(node->right);
            }
            // go to next level
            level++;
        }
        return levels;
    }
};
